### DISCUSSION TITLE
Use of `typing.Self` in `__init__`

### DISCUSSION THREAD

**Discussion by tpgillam** [microsoft/pyright#8531] [General]
Consider the following example:

```python
import typing

class A:
    def __init__(self: typing.Self) -> None:
        pass
```

Pyright will flag a `reportInvalidTypeVarUse` over the use of `Self` here. However, by my reading of the recent specification, this (probably?) ought to be allowed:

https://github.com/python/typing/blob/b908de0205fb208cd2d74fe342ab33f28f35b71b/docs/spec/constructors.rst?plain=1#L206-L207

> If the ``self`` parameter within the ``__init__`` method is not annotated, type checkers should infer a type of ``Self``.

I am curious about what the correct resolution is for the inconsistency!

Also, in _practice_ I am guessing that the general consensus is to discourage explicit annotation of `self` in general? e.g. I note that Ruff's [ANN01](https://docs.astral.sh/ruff/rules/missing-type-self/) rule is deprecated.

**Comment by erictraut** [comment]
Pyright should not emit a diagnostic in this case. I'm going to convert this to a bug report.

And yes, there is no need to annotate `self` in general. Its type is implicitly `Self` unless you want to override this default. Likewise, the `cls` parameter in a class method always has the implicit type `type[Self]` unless explicitly overridden.

**Reply by erictraut** [reply]
Here's a link to the bug report: https://github.com/microsoft/pyright/issues/8532

**Reply by tpgillam** [reply]
Thanks very much for addressing this so quickly!
